### ISSUE TITLE
SIGSEGV when when mock with shared_ptr

### ISSUE THREAD

**Issue by zhlongfj** [eranpeer/FakeIt#286]
My code:
auto mock1 = new Mock<IMeta>();
When(Method(*mock1, resId)).AlwaysReturn("1");
auto meta = shared_ptr<IMeta>(&mock1->get());

error info:
FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal
error in "method(c);" (line: 5541)

**Comment by FranckRJ** [comment]
Don't create the mock on the heap, create it on the stack. And you need to specialize it otherwise it doesn't work.

**Comment by zhlongfj** [comment]
That means I can't use shared_ptr to wrap the point from mock().get()?

**Comment by FranckRJ** [comment]
You can, delete operator is overrided so it does nothing.

**Comment by malcolmdavey** [comment]
Yes override the delete operator

```
My code:
Mock mock1;
When(Method(mock1, resId)).AlwaysReturn("1");
auto meta = shared_ptr(mock1->get(), [](auto){});
```

**Comment by malcolmdavey** [comment]
Might be useful having this in the docs, though.

**Comment by FranckRJ** [comment]
What i mean is that you don't need to do anything. It works without any workaround.

**Comment by FranckRJ** [comment]
```
Mock mock1;
When(Method(mock1, resId)).AlwaysReturn("1");
auto meta = shared_ptr(mock1.get());
```

It's enough.

**Comment by zhlongfj** [comment]
> Yes override the delete operator
> 
> ```
> My code:
> Mock mock1;
> When(Method(mock1, resId)).AlwaysReturn("1");
> auto meta = shared_ptr(mock1->get(), [](auto){});
> ```

It works, thanks

**Comment by zhlongfj** [comment]
> You can, delete operator is overrided so it does nothing.

It works, thanks

**Comment by FranckRJ** [comment]
I've looked a bit more into the source code to understand how it works, the summary (for GCC) is:
- there isn't any overrided delete operator
- GCC has in fact two entries for the destructor in the VTable, the first one point to the "~Class()" provided by the user. The second one call the first one and the destructor of any parent classes as well, at the end it also release the memory (free).
- When a mock is created, every entries in the VTable is set to a function that throws "method not implemented" except for the two destructors that are set to a method that does nothing.
- When the smart pointer is destroyed and call "delete" on the object it holds, it call the second destructor that does nothing, it doesn't even call free(). That's why later when the Mock object is destroyed because it goes out of scope it doesn't throw a "double free" error or something like that, because the object was never destroyed.

For visual studio it's a bit the same but instead of two destructors in the vtable there is only one that point to a wrapper that call "~Class()", etc (and free).

In the end, what you need to remember is that you have nothing special to do, use the pointer without worrying about any potential double free. You don't need to set any custom deleter / destructor.

**Comment by FranckRJ** [comment]
And also I wasn't able to reproduce the crash with a mock allocated on the heap : https://godbolt.org/z/dcTosana8

I mean, it isn't necessary so you shouldn't do it anyway but I still wonder why it crashed for you.

**Comment by zhlongfj** [comment]
the follow code is crashed in my VS2019(C++17), but It's OK in: https://godbolt.org/z/dcTosana8

class Demo
{
public:
    shared_ptr<IMeta> _meta;
};


void testcase()
{
    Mock<IMeta> mock2;
    When(Method(mock2, resId)).AlwaysReturn("1");
    Demo demo;
    demo._meta = shared_ptr<IMeta>(&mock2.get());
}

crash info:
0x0000000000000000 Exception raised at TimelineUnitTest.exe: 0xC0000005: Access conflict occurred at execution location 0x0000000000000000.

**Comment by malcolmdavey** [comment]
I'll double check tomorrow on my machine - also with VS 2019.

**Comment by malcolmdavey** [comment]
I've also found I need to do this on VS 2019/v142.
There may be a bug with it on VS

**Comment by FranckRJ** [comment]
Fixed by #289 in FakeIt 2.3.1.
